Added support for the Adafruit i2c/SPI LCD Backpack

GoranLundberg opened the pull request:
Added support for the Adafruit i2c/SPI LCD Backpack
https://learn.adafruit.com/i2c-spi-lcd-backpack/overview

The MCP23008 uses registers so the code had to be altered a bit.

From i2c.py:
```

      CharLCD via Adafruit i2c/SPI LCD Backback with MCP23008 I2C port expander:

            Warning: You will most likely need a level shifter (that supports i2c)
            between the SCL/SDA connections on the backpack and the Raspberry Pi.
            Or you might damage the Pi and possibly any other 3.3V i2c devices
            connected on the i2c bus. The SCL/SDA outputs are rated 0.7*VDD
            which means the MCP23008 outputs 3.5V when 5V is applied to drive the LCD.

            Will most likely work with only a MCP23008 as well, but not tested.
            The MCP23008 needs to be connected the exactly the same way as the backpack.

            For complete schematics see the adafruit page at:
            https://learn.adafruit.com/i2c-spi-lcd-backpack/

            4-bit operation only supported.

            Pin mapping::

            7    |  6  |  5  |  4  |  3  | 2 | 1  | 0
            LITE | DB7 | DB6 | DB5 | DB4 | E | RS | -

            Note: MCP23017 is NOT supported, but the code can easily be adapted for it.
```

**Code issues:**

**1.**
The MCP code is using _pulse_data() directly instead of using _write4bits() as the PCF is doing.
It's a little bit messy, but it seemed wierd to to call _write4bits() and add a if statement for the different chips, and just calling _pulse_data() and do nothing else for the MCP in the _write4bit() func.
Any idea to make this code cleaner?  What do you prefer?

Edit: Changing the write4bit function is more consistent.
```

    def _send(self, value, mode):
        """Send the specified value to the display with automatic 4bit / 8bit selection.
        The rs_mode is either ``common.RS_DATA`` or ``common.RS_INSTRUCTION``."""
        if(self.i2c_expander == "PCF8574"):
            self._write4bits(mode | (value & 0xF0))
            self._write4bits(mode | ((value << 4) & 0xF0))
        elif(self.i2c_expander == "MCP23008"):
            if(mode == c.RS_INSTRUCTION):
                self._mcp_data &= ~c.MCP23008_RS
                self._pulse_data(value >> 4)
                self._pulse_data(value & 0x0F)
            elif(mode == c.RS_DATA):
                self._mcp_data |= c.MCP23008_RS
                self._pulse_data(value >> 4)
                self._pulse_data(value & 0x0F)
        else:
            raise NotImplementedError('I2C expander is not supported.')

# now its like this. but only called in the PCF code.

    def _write4bits(self, value):
        """Write 4 bits of data into the data bus."""
        self.bus.write_byte(self.address, value | self._backlight)
        self._pulse_data(value)

# as opposed to:

    def _write4bits(self, value):
        """Write 4 bits of data into the data bus."""
        if(self.i2c_expander == "PCF8574"):
             self.bus.write_byte(self.address, value | self._backlight)
             self._pulse_data(value)
       elif(self.i2c_expander == "MCP23008"):
             self._pulse_data(value)

```

**2.**
The _send() function (as below) is a bit messy. I suggest separating it to two different functions.
_send_data() and  _send_command() or similar. Unless you have a reason to do it like this.
```
 def _send(self, value, mode):
        """Send the specified value to the display with automatic 4bit / 8bit selection.
        The rs_mode is either ``common.RS_DATA`` or ``common.RS_INSTRUCTION``."""
        if(self.i2c_expander == "PCF8574"):
            self._write4bits(mode | (value & 0xF0))
            self._write4bits(mode | ((value << 4) & 0xF0))
        elif(self.i2c_expander == "MCP23008"):
            if(mode == c.RS_INSTRUCTION):
                self._mcp_data &= ~c.MCP23008_RS
                self._pulse_data(value >> 4)
                self._pulse_data(value & 0x0F)
            elif(mode == c.RS_DATA):
                self._mcp_data |= c.MCP23008_RS
                self._pulse_data(value >> 4)
                self._pulse_data(value & 0x0F)
```


I'm planning to add support for the MCP23017 as well when I get a chance.

If you have any questions please ask. If you have some improvements to my code I'm all ears.

**Note: I don't have a PCF8572, so I have not tested the code on that expander. Needs to be tested before added to master branch. The MCP23008 Backpack is working perfectly though.**

dbrgn commented:
Thanks! :) I'll take a look probably later this week.

I don't have the backpack myself, but it looks like the shop where I usually buy my electronics stuff has become an Adafruit distributor, so maybe I'll get myself one. (Shipping from Adafruit directly to Switzerland is expensive.)

GoranLundberg commented:
I incorporated it into i2c.py but maybe creating a separate file/subclass for each chipset is better. Code will be cleaner, but maintaining i2c support in one file for each chipset will be harder.

Depends which route you want to go. I can rewrite it in a separate file if you prefer that. Just let me know.

dbrgn commented:
I just ordered one of those backpacks for testing.

dbrgn commented on RPLCD/i2c.py:
```diff
@@ -77,7 +110,16 @@ def __init__(self, address, port=1,
+            c.msleep(50)
+        elif(self.i2c_expander == "MCP23008"):
+            # Set IO DIRection to output on all GPIOs (GP0-GP7)
+            self.bus.write_byte_data(self.address, c.MCP23008_IODIR, 0x00)
```
Is `0x00` the constant for "output"? If yes, maybe put it into a variable called `MCP23008_IODIR_OUT` or `MCP23008_MODE_OUT` or something like that?

GoranLundberg replied:
Its not a constant. 0x00 is sent to the IODIR register, one bit for each gpio on the expander.
For example sending 0b10000010 would set gpio 7 and 1 to input (1) and the rest to output (0).
Don't think that setting it to a constant would make it easier to read the code. It's only used once and will not change.

dbrgn replied:
Ah, I see. I think it's fine :)

dbrgn commented:
For the record, I'll release version 0.9 (as release candidate for stabilization) soon, but maybe we can land the MCP23008 support before 1.0 :)

GoranLundberg commented on RPLCD/common.py:
```diff
@@ -84,6 +88,25 @@
+MCP23008_INTCAP = 0x08  # (Read-only)
+MCP23008_GPIO = 0x09
+MCP23008_OLAT = 0x0A
+
```
Will try to rebase with the main branch. And make a new pull request with the new changes.
I'll probably order one of the PCF expanders. They are quite cheap.

dbrgn replied:
You don't have to do a new pull request, you can use "git push -f" to overwrite the old branch.

GoranLundberg commented on RPLCD/i2c.py:
```diff
@@ -57,18 +82,26 @@ def __init__(self, address, port=1,
-
+        if(self.i2c_expander == "PCF8574"):
+            self._backlight = c.LCD_BACKLIGHT if backlight_enabled else c.LCD_NOBACKLIGHT
+        elif(self.i2c_expander == "MCP23008"):
```
:) old C habits. Will fix it.

dbrgn replied:
No worries :)

GoranLundberg commented on RPLCD/i2c.py:
```diff
@@ -57,18 +82,26 @@ def __init__(self, address, port=1,
+        elif(self.i2c_expander == "MCP23008"):
+            self._backlight = c.MCP23008_BACKLIGHT if backlight_enabled else c.MCP23008_NOBACKLIGHT
+        else:
+            raise NotImplementedError('I2C expander is not supported.')
```
Absolutely.  But I'm not sure all of them are needed. Probably better to do it once when initiated.
And will rename i2c_expander to _i2c_expander since it's a really bad idea to let users change it after the class is initiated.

dbrgn commented on RPLCD/i2c.py:
```diff
@@ -32,15 +32,40 @@ class CharLCD(BaseCharLCD):
                        cols=20, rows=4, dotsize=8,
                        auto_linebreaks=True,
-                       backlight_enabled=True):
+                       backlight_enabled=True, i2c_expander="PCF8574"):
```
Hm, you're right. I should have merged this before the PyPI release, then we wouldn't have any reverse compatibility issues.

On the other hand the PCF8574 is quite widely used, so it's probably a good default.

dbrgn commented on RPLCD/gpio.py:
```diff
@@ -156,12 +156,30 @@ def _set_backlight_enabled(self, value):
-    def _send(self, value, mode):
-        """Send the specified value to the display with automatic 4bit / 8bit
-        selection. The rs_mode is either ``RS_DATA`` or ``RS_INSTRUCTION``."""
+    def _send_data(self, value):
```
Hm, I think in the case of the GPIO version, having a single version with a mode parameter might probably be better, since the method body is almost completely the same. Maybe you can re-create the `_send` method and call it from `_send_data` and `_send_instruction`?

dbrgn commented on RPLCD/i2c.py:
```diff
@@ -61,18 +109,26 @@ def __init__(self, address, port=1,
         self.port = port
 
+        # Set i2c expander, "PCF8574" and "MCP23008" are supported.
+        self._i2c_expander = i2c_expander
```
Maybe you could move the expander check from line 130 to here? Makes more sense to do it when assigning the variable, instead of waiting for backlight control to check whether the value is valid.

Something like this:

```python
if i2c_expander in ['PCF8574', 'MCP23008']:
    self._i2c_expander  = i2c_expander
else:
    raise NotImplementedError('I2C expander "%s" is not supported.' % i2c_expander)
```

dbrgn commented:
> Did not change the new tests. I'm not familiar with pytest or mocker. Needs to be fixed.

I'll push a fix to your branch.

dbrgn commented:
I just tested your branch with the PCF8574, works fine :)

dbrgn commented:
Ah, by the way, in case you have questions or something else to discuss, I created a new Gitter channel yesterday: https://gitter.im/RPLCD/Lobby

GoranLundberg commented:
- Changed gpio.py as suggested
- Changed error handling for i2c_expander.
- PEP8 Errors fixed (I think)
- In i2c.py added underscore to the following variables. Don't see a reason for users to modify these.
      self._address = address
       self._port = port
- Minor modifications to info about MCP23008

dbrgn commented:
Great! Looks ready to merge now, as soon as I found time to test :) Probably this weekend.
